### PR TITLE
Better rewards + refactoring

### PR DESCRIPTION
**Pull request by andrgolubev**

### CODE REVIEW

File: madcars/runners/localrunner.py [hunk 1]
```diff
@@ -7,12 +7,12 @@
 import os
 import sys
 
-maps = ['PillMap']#,'PillHubbleMap', 'PillHillMap', 'PillCarcassMap', 'IslandMap', 'IslandHoleMap']
-cars = ['Buggy']#, 'Bus', 'SquareWheelsBuggy']
```
Comment (Wovchena) [inline madcars/runners/localrunner.py]: The idea was to narrow a state space to make it easier to train. Can you train the agent well on at least one map?

File: madcars/players/pytorch_main.py [hunk 2]
```diff
@@ -141,17 +144,19 @@ def compute_td_loss(states, actions, rewards, next_states, is_done, gamma=0.99,
...
-                    wines+=1
+                    rewards[-1] += 100
+                    wins+=1
                 else:
-                    rewards[-1] = -1
+                    rewards[-1] += -100
                     loses+=1
                 prev_lives=lives
                 if VERBOSE and gamecount%20==0:
+                    # FI.write(z)
```
Comment (Wovchena) [inline madcars/players/pytorch_main.py]: What is z?) remove it?

File: madcars/players/pytorch_main.py [hunk 3]
```diff
@@ -207,6 +219,8 @@ def compute_td_loss(states, actions, rewards, next_states, is_done, gamma=0.99,
             states.append(state)
             actions.append(choice)
             rewards.append(0)
+            for i in range(len(rewards)):
+                rewards[i] -= 10  # penalize for time spent not winning
```
Comment (Wovchena) [inline madcars/players/pytorch_main.py]: It looks that the longer the sequence the smaller the reward. So it requires agent to predict a game duration to predict the reward. For the reward of the first ticks it sounds weird. Does it work for you? The agent learns nothing for me.

File: madcars/runners/localrunner.py [hunk 4]
```diff
@@ -7,12 +7,12 @@
...
-games = [','.join(t) for t in product(maps, cars)]*101
-cur_dir = os.path.dirname(os.path.basename(__file__))
+maps = ['PillMap','PillHubbleMap', 'PillHillMap', 'PillCarcassMap', 'IslandMap', 'IslandHoleMap']
+cars = ['Buggy', 'Bus', 'SquareWheelsBuggy']
+games = [','.join(t) for t in product(maps, cars)]*10
```
Comment (Wovchena) [inline madcars/runners/localrunner.py]: I suggest having 2 games only, so the agent could always play vs itself. In the current setup it plays with not updating enemy so the sequences of actions will be almost the same. It is like having bigger learning rate.

File: madcars/players/pytorch_main.py [hunk 5]
```diff
@@ -207,6 +219,8 @@ def compute_td_loss(states, actions, rewards, next_states, is_done, gamma=0.99,
...
             dones.append(False)
             # update epsilon:
             epsilon = min(epsilon * epsilon_decay, 0.1)
```
Comment (Wovchena) [inline madcars/players/pytorch_main.py]: (Should there be max?)
There is no reason to decay epsilon in local memory because after 10 games it will be reset. I believe you never even get to 0.1.

File: madcars/players/pytorch_main.py (hunk 2 above)
Comment (Wovchena) [reply madcars/players/pytorch_main.py]: @andrgolubev

**Review by Wovchena — COMMENTED** [review]
kek

File: madcars/runners/localrunner.py (hunk 1 above)
Comment (andrgolubev) [reply madcars/runners/localrunner.py]: I do. Training should work well on all maps (even on a single one). I have a preference towards learning at all maps/all cars in order to make agent weights more generic.
Otherwise, what I saw is that agent works properly only for a single environment

File: madcars/players/pytorch_main.py (hunk 2 above)
Comment (andrgolubev) [reply madcars/players/pytorch_main.py]: some debugging artifacts

File: madcars/players/pytorch_main.py (hunk 3 above)
Comment (andrgolubev) [reply madcars/players/pytorch_main.py]: Oh, you're actually right. I think there's a **bug** here. What I meant is to have a way to penalize reward  in time. It would be cool to be able to accumulate the penalty over-time however, but I think it could be achieved much simpler:
~~~
rewards[-1] -= 10
~~~
This should work OK since we have a Q-value that is concerned about previous rewards (although, multiplied by `gamma`)

What's you opinon @Wovchena ?

File: madcars/runners/localrunner.py (hunk 4 above)
Comment (andrgolubev) [reply madcars/runners/localrunner.py]: Makes sense. But let's instead do weight sharing to fix this. Leaning towards having multiple maps. Not a blocker for me, though, can revert back

File: madcars/players/pytorch_main.py (hunk 5 above)
Comment (andrgolubev) [reply madcars/players/pytorch_main.py]: there should be **max**, right, will fix.

let's leave it "as is" for now. in the long run we should be able to benefit from it (right?)

**Review by andrgolubev — COMMENTED** [review]
see per-comment answers

File: madcars/runners/localrunner.py (hunk 1 above)
Comment (Wovchena) [reply madcars/runners/localrunner.py]: The pushed model does suicide.

File: madcars/runners/localrunner.py (hunk 4 above)
Comment (Wovchena) [reply madcars/runners/localrunner.py]: If you you are going to have weight sharing you probably want to have 100 games, as it was earlier. I do not see the reason to have 10. It should be 2 or the big number, but not something in the middle

File: madcars/players/pytorch_main.py (hunk 3 above)
Comment (Wovchena) [reply madcars/players/pytorch_main.py]: You can simply do `rewards.append(-10)`

File: madcars/players/pytorch_main.py (hunk 5 above)
Comment (Wovchena) [reply madcars/players/pytorch_main.py]: Okay (may be)

File: madcars/runners/run_original_env.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+echo "------------------------------------"
+echo "agent location: $agent_folder"
+echo "runner location: $runner_folder"
+echo "sessions number: $num_runs"
+echo "------------------------------------"
+echo "Game:"
+
+for ((i=0; i < $num_runs; ++i));
+do
+    python3 $runner_folder/localrunner.py -f "python3 $agent_folder/pytorch_main.py --train" -s "python3 $agent_folder/pytorch_main.py"
```
Comment (Wovchena) [inline madcars/runners/run_original_env.sh]: btw, it is advised to have a -u key

Comment (andrgolubev) [reply madcars/runners/run_original_env.sh]: `-u` is for unbuffered python io. we don't really do any io to the user. I don't think it's necessary. This might cause troubles for localrunner.py reading the output though. Good catch

File: madcars/runners/localrunner.py (hunk 4 above)
Comment (andrgolubev) [reply madcars/runners/localrunner.py]: This is actually a big number, 10 games multiplied by a product of maps and cars

**Comment by andrgolubev** [comment]
Thought about one more feature: do weight sharing and train agents on different games (e.g. `agent1` trains every **even** game, `agent2` trains every **non even** game)

**Comment by Wovchena** [comment]
How will you merge both trained agents?

File: madcars/runners/localrunner.py (hunk 4 above)
Comment (Wovchena) [reply madcars/runners/localrunner.py]: you could have bigger, but ok, up to you

**Comment by andrgolubev** [comment]
I already do this but they both train in the same game: via `agent.pth` and file locking for interprocess communication. By the way, please check that the approach I use work of Windows. ~It seems like it's `POSIX` api, so no idea how Windows python lib is implemented~

**Comment by Wovchena** [comment]
It does not look that you do agents merging. It is more like you share the same agent weights between two player.
This allows agent to view at situation from two points of view and it is probably better then training every even/not even game as the second approach reduces the number oа samples by 2 times.

**Comment by andrgolubev** [comment]
I am not sure what's agent merging. I understand weights sharing as a use of the same network weights that is trained simultaneously from two directions. This is what I implemented at least, am I wrong?

**Comment by andrgolubev** [comment]
Okay, I guess I don't understand what you were asking at the beginning:

> How will you merge both trained agents?

From my perspective: we train two agents with shared weights, then we use trained agent when playing in "production". There's a single weights blob, so there's no "merged" agents. Instead, this is one agent from all points of view, IMHO

**Comment by Wovchena** [comment]
Then you should omit this messages:

> How will you merge both trained agents?

and

> It does not look that you do agents merging. It is more like you share the same agent weights between two player.

The answer for the question

> Thought about one more feature: do weight sharing and train agents on different games (e.g. agent1 trains every even game, agent2 trains every non even game)

is

> This allows agent to view at situation from two points of view and it is probably better then training every even/not even game as the second approach reduces the number oа samples by 2 times.
